File: frontend/src/utils/realtimeGuard.js
const streamState = new Map();

function toSequenceNumber(value) {
  if (value === null || value === undefined) return null;
  if (typeof value === 'number' && Number.isFinite(value)) return value;
  if (typeof value === 'string' && value.trim() !== '') {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return parsed;
  }
  return null;
}

function readSequence(event) {
  return (
    toSequenceNumber(event?.sequence)
    ?? toSequenceNumber(event?.seq)
    ?? toSequenceNumber(event?.eventSequence)
    ?? toSequenceNumber(event?.payload?.sequence)
    ?? toSequenceNumber(event?.payload?.seq)
    ?? toSequenceNumber(event?.payload?.eventSequence)
  );
}

export function evaluateRealtimeEvent(streamKey, eventType, event) {
  const sequence = readSequence(event);
  if (sequence === null) {
    return { gapDetected: false, sequence: null, expected: null };
  }

  const stateKey = `${streamKey}:${eventType || 'unknown'}`;
  const previous = streamState.get(stateKey);
  const expected = previous === undefined ? null : previous + 1;
  const gapDetected = expected !== null && sequence > expected;

  if (previous === undefined || sequence > previous) {
    streamState.set(stateKey, sequence);
  }

  return { gapDetected, sequence, expected };
}

export function resetRealtimeStream(streamKey) {
  for (const key of Array.from(streamState.keys())) {
    if (key.startsWith(`${streamKey}:`)) {
      streamState.delete(key);
    }
  }
}

File: mobile/lib/src/core/utils/realtime_guard.dart
class RealtimeGuardResult {
  final bool gapDetected;
  final int? sequence;
  final int? expected;
  const RealtimeGuardResult({
    required this.gapDetected,
    this.sequence,
    this.expected,
  });
}

class RealtimeGuard {
  static final Map<String, int> _streamState = {};

  static int? _toSequence(dynamic value) {
    if (value == null) return null;
    if (value is int) return value;
    if (value is String && value.trim().isNotEmpty) {
      return int.tryParse(value);
    }
    return null;
  }

  static int? _readSequence(Map<String, dynamic> event) {
    return _toSequence(event['sequence']) ??
        _toSequence(event['seq']) ??
        _toSequence(event['eventSequence']) ??
        _toSequence((event['payload'] as Map?)?['sequence']) ??
        _toSequence((event['payload'] as Map?)?['seq']);
  }

  static RealtimeGuardResult evaluate(String streamKey, String? eventType, Map<String, dynamic> event) {
    final sequence = _readSequence(event);
    if (sequence == null) {
      return const RealtimeGuardResult(gapDetected: false);
    }

    final stateKey = '$streamKey:${eventType ?? 'unknown'}';
    final previous = _streamState[stateKey];
    final expected = previous == null ? null : previous + 1;
    final gapDetected = expected != null && sequence > expected;

    if (previous == null || sequence > previous) {
      _streamState[stateKey] = sequence;
    }

    return RealtimeGuardResult(gapDetected: gapDetected, sequence: sequence, expected: expected);
  }

  static void resetStream(String streamKey) {
    _streamState.removeWhere((key, _) => key.startsWith('$streamKey:'));
  }
}
